File: phases/07-transformers-deep-dive/04-positional-encoding/code/main.jl
# 用 Julia 实现位置编码：正弦绝对位置、旋转位置嵌入（RoPE）和 ALiBi
# 偏置矩阵。验证 RoPE 点积仅取决于相对距离。仅使用标准库。资料来源：
#   https://arxiv.org/abs/2104.09864
#   https://arxiv.org/abs/2108.12409
#   https://docs.julialang.org/en/v1/manual/mathematical-operations/

using Random
using Printf


function sinusoidal_pe(n::Int, d::Int; base::Float64=10000.0)::Matrix{Float64}
    n > 0 || throw(ArgumentError("n 必须大于 0"))
    d > 0 || throw(ArgumentError("d 必须大于 0"))
    iseven(d) || throw(ArgumentError("d 必须为偶数，以便组成正弦/余弦对"))
    pe = zeros(n, d)
    for pos in 0:(n - 1)
        for i in 0:(d ÷ 2 - 1)
            theta = pos / (base ^ (2 * i / d))
            pe[pos + 1, 2 * i + 1] = sin(theta)
            pe[pos + 1, 2 * i + 2] = cos(theta)
        end
    end
    return pe
end


function apply_rope(x::Vector{Float64}, pos::Int; base::Float64=10000.0)::Vector{Float64}
    d = length(x)
    iseven(d) || throw(ArgumentError("RoPE 要求嵌入维度为偶数"))
    out = copy(x)
    for i in 0:(d ÷ 2 - 1)
        theta = pos / (base ^ (2 * i / d))
        c = cos(theta)
        s = sin(theta)
        a = x[2 * i + 1]
        b = x[2 * i + 2]
        out[2 * i + 1] = a * c - b * s
        out[2 * i + 2] = a * s + b * c
    end
    return out
end


function dotprod(a::Vector{Float64}, b::Vector{Float64})::Float64
    return sum(a .* b)
end


function alibi_slopes(n_heads::Int)::Vector{Float64}
    n_heads > 0 || throw(ArgumentError("n_heads 必须大于 0"))
    return [2.0 ^ (-8.0 * (h) / n_heads) for h in 1:n_heads]
end


function alibi_bias(n_heads::Int, seq_len::Int; causal::Bool=true)
    slopes = alibi_slopes(n_heads)
    out = Vector{Matrix{Float64}}()
    for m in slopes
        bias = fill(0.0, seq_len, seq_len)
        for i in 1:seq_len
            for j in 1:seq_len
                if causal && j > i
                    bias[i, j] = -Inf
                else
                    bias[i, j] = -m * abs(i - j)
                end
            end
        end
        push!(out, bias)
    end
    return out
end


function demo_sinusoidal()
    println("=== 正弦位置编码 ===")
    pe = sinusoidal_pe(8, 8)
    println("前 4 个位置、前 4 个维度：")
    for pos in 1:4
        row_str = join([@sprintf("%+.3f", pe[pos, j]) for j in 1:4], "  ")
        @printf("  pos=%d: %s\n", pos - 1, row_str)
    end
    println()
end


function demo_rope_relative()
    println("=== RoPE：点积仅取决于相对距离 ===")
    rng = MersenneTwister(0)
    d = 16
    q = randn(rng, d)
    k = randn(rng, d)
    pairs = [(3, 5), (7, 9), (100, 102), (1024, 1026)]
    @printf("%6s  %6s  %4s  %18s\n", "pos_q", "pos_k", "间隔", "<q_rot, k_rot>")
    for (pq, pk) in pairs
        q_rot = apply_rope(q, pq)
        k_rot = apply_rope(k, pk)
        d_prod = dotprod(q_rot, k_rot)
        @printf("%6d  %6d  %4d  %18.6f\n", pq, pk, pk - pq, d_prod)
    end
    println("间隔为 2 的所有行都应产生相同点积。")
    println()
end


function demo_rope_base_scaling()
    println("=== RoPE 基数缩放（面向长上下文的 NTK-aware 方法）===")
    rng = MersenneTwister(1)
    d = 8
    q = randn(rng, d)
    k = randn(rng, d)
    for base in (10000.0, 100000.0, 1_000_000.0)
        q_rot = apply_rope(q, 4096; base=base)
        k_rot = apply_rope(k, 4098; base=base)
        @printf("  基数=%8d  得分=%+.6f\n", Int(base), dotprod(q_rot, k_rot))
    end
    println("基数越大 = 旋转越慢 = 不发生相位环绕的上下文越长。")
    println()
end


function demo_alibi()
    println("=== ALiBi 偏置矩阵 ===")
    n_heads = 4
    slopes = alibi_slopes(n_heads)
    @printf("%d 个头的斜率：%s\n", n_heads,
            join([@sprintf("%.4f", s) for s in slopes], ", "))
    bias = alibi_bias(n_heads, 6; causal=false)
    println("第 1 个头的偏置（token 越近，惩罚越小）：")
    for row in eachrow(bias[1])
        println("  " * join([@sprintf("%+6.2f", v) for v in row], "  "))
    end
    println()
end


function main()
    demo_sinusoidal()
    demo_rope_relative()
    demo_rope_base_scaling()
    demo_alibi()
    println("要点：RoPE 在点积内部编码相对位置；")
    println("ALiBi 完全跳过嵌入。如今正弦位置编码已退居次要位置。")
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: phases/07-transformers-deep-dive/04-positional-encoding/code/main.rs
// 位置编码：正弦编码、RoPE、ALiBi。仅使用标准库。
// 主题：将 token 位置编码到查询、键或注意力偏置中。
// 参考资料（仅作为引用，不作为依赖）：
//   - Vaswani 2017（正弦编码）：https://arxiv.org/abs/1706.03762
//   - Su 等（2021，RoPE）：https://arxiv.org/abs/2104.09864
//   - Press 等（2021，ALiBi）：https://arxiv.org/abs/2108.12409
//   - candle 的 RoPE 实现：https://github.com/huggingface/candle/blob/main/candle-nn/src/rotary_emb.rs
//
// 编译并运行：rustc --edition 2021 main.rs -o /tmp/pe && /tmp/pe

use std::f32::consts::PI;

// 正弦位置编码表 [n, d]。
fn sinusoidal_pe(n: usize, d: usize, base: f32) -> Vec<Vec<f32>> {
    let mut pe = vec![vec![0.0f32; d]; n];
    for pos in 0..n {
        for i in 0..(d / 2) {
            let theta = (pos as f32) / base.powf(2.0 * i as f32 / d as f32);
            pe[pos][2 * i] = theta.sin();
            pe[pos][2 * i + 1] = theta.cos();
        }
    }
    pe
}

// 将 x 的奇偶元素对旋转 pos * theta_i 角度，返回新的 Vec。
fn apply_rope(x: &[f32], pos: usize, base: f32) -> Vec<f32> {
    let d = x.len();
    let mut out = x.to_vec();
    for i in 0..(d / 2) {
        let theta = (pos as f32) / base.powf(2.0 * i as f32 / d as f32);
        let c = theta.cos();
        let s = theta.sin();
        let a = x[2 * i];
        let b = x[2 * i + 1];
        out[2 * i] = a * c - b * s;
        out[2 * i + 1] = a * s + b * c;
    }
    out
}

fn dot(a: &[f32], b: &[f32]) -> f32 {
    a.iter().zip(b.iter()).map(|(x, y)| x * y).sum()
}

// ALiBi 斜率：当 h 位于 0..n_heads 时为 2^(-8*(h+1)/n_heads)。
fn alibi_slopes(n_heads: usize) -> Vec<f32> {
    (0..n_heads)
        .map(|h| 2.0f32.powf(-8.0 * (h + 1) as f32 / n_heads as f32))
        .collect()
}

// 各个头的 ALiBi 偏置矩阵：-slope * |i - j|，可选因果掩码。
fn alibi_bias(n_heads: usize, seq_len: usize, causal: bool) -> Vec<Vec<Vec<f32>>> {
    let slopes = alibi_slopes(n_heads);
    let mut out = Vec::with_capacity(n_heads);
    for &m in &slopes {
        let mut head = vec![vec![0.0f32; seq_len]; seq_len];
        for i in 0..seq_len {
            for j in 0..seq_len {
                head[i][j] = if causal && j > i {
                    f32::NEG_INFINITY
                } else {
                    -m * (i as i64 - j as i64).abs() as f32
                };
            }
        }
        out.push(head);
    }
    out
}

// 用于生成确定性高斯样本的微型 LCG。
struct Rng { state: u64 }
impl Rng {
    fn new(seed: u64) -> Self { Rng { state: seed.wrapping_mul(0x9E37_79B9_7F4A_7C15) | 1 } }
    fn next_u32(&mut self) -> u32 {
        self.state = self.state.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        (self.state >> 33) as u32
    }
    fn uniform(&mut self) -> f32 { (self.next_u32() as f32 + 1.0) / (u32::MAX as f32 + 2.0) }
    fn gauss(&mut self) -> f32 {
        let u1 = self.uniform();
        let u2 = self.uniform();
        (-2.0 * u1.ln()).sqrt() * (2.0 * PI * u2).cos()
    }
}

fn demo_sinusoidal() {
    println!("=== 正弦位置编码 ===");
    let pe = sinusoidal_pe(8, 8, 10000.0);
    println!("前 4 个位置、前 4 个维度：");
    for pos in 0..4 {
        print!("  pos={}: ", pos);
        for i in 0..4 {
            print!(" {:+.3}", pe[pos][i]);
        }
        println!();
    }
    println!();
}

fn demo_rope_relative() {
    println!("=== RoPE：点积仅取决于相对距离 ===");
    let mut rng = Rng::new(0);
    let d = 16usize;
    let q: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();
    let k: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();

    let pairs = [(3usize, 5usize), (7, 9), (100, 102), (1024, 1026)];
    println!(" {:>6}  {:>6}  {:>4}  {:>18}", "pos_q", "pos_k", "间隔", "<q_rot, k_rot>");
    for (pq, pk) in pairs {
        let q_rot = apply_rope(&q, pq, 10000.0);
        let k_rot = apply_rope(&k, pk, 10000.0);
        let d_prod = dot(&q_rot, &k_rot);
        println!(" {:>6}  {:>6}  {:>4}  {:>18.6}", pq, pk, (pk as i64) - (pq as i64), d_prod);
    }
    println!("间隔为 2 的所有行都具有相同点积。");
    println!();
}

fn demo_rope_base_scaling() {
    println!("=== RoPE 基数缩放（面向长上下文的 NTK-aware 方法）===");
    let mut rng = Rng::new(1);
    let d = 8usize;
    let q: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();
    let k: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();

    for base in [10_000.0f32, 100_000.0, 1_000_000.0] {
        let q_rot = apply_rope(&q, 4096, base);
        let k_rot = apply_rope(&k, 4098, base);
        println!("  基数={:>9}  得分={:+.6}", base as u64, dot(&q_rot, &k_rot));
    }
    println!("基数越大 = 旋转越慢 = 不发生相位环绕的上下文越长。");
    println!();
}

fn demo_alibi() {
    println!("=== ALiBi 偏置矩阵 ===");
    let n_heads = 4usize;
    let slopes = alibi_slopes(n_heads);
    print!("{} 个头的斜率：", n_heads);
    for s in &slopes { print!(" {:.4}", s); }
    println!();
    let bias = alibi_bias(n_heads, 6, false);
    println!("第 0 个头的偏置（token 越近，惩罚越小）：");
    for row in &bias[0] {
        print!(" ");
        for v in row { print!(" {:+6.2}", v); }
        println!();
    }
    println!();
}

fn demo_rope_microbench() {
    println!("=== 微基准测试：5 万次 RoPE 旋转（d=128）===");
    let mut rng = Rng::new(2);
    let d = 128usize;
    let q: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();
    let start = std::time::Instant::now();
    let mut sink = 0.0f32;
    for pos in 0..50_000usize {
        let r = apply_rope(&q, pos, 10_000.0);
        sink += r[0];
    }
    let elapsed = start.elapsed();
    println!("5 万次旋转耗时 {:.2}ms（{:.0}/秒）  累加值={:.4}",
        elapsed.as_secs_f64() * 1000.0,
        50_000.0 / elapsed.as_secs_f64(),
        sink,
    );
}

fn main() {
    demo_sinusoidal();
    demo_rope_relative();
    demo_rope_base_scaling();
    demo_alibi();
    demo_rope_microbench();
    println!();
    println!("要点：RoPE 在点积本身中编码相对位置。");
    println!("ALiBi 完全跳过嵌入。如今正弦位置编码大多只具有历史意义。");
}
